codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -111,123 +112,152 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
             int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
             verifyPayloadLength(payloadLength);
 
-            ByteBuf out = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
+            writeFrameHeader(ctx, payloadLength, DATA, flags, streamId, promiseAggregator.newPromise());
```

Comment:
Yes that it though it requires cooperation with the allocator for it to be effective when we're talking about DATA frames.

If the layer above HTTP2 does

Buffer myPayload = ctx().alloc().buffer(100);
... write some stuff to myPayload....
http2Encoder.writeData(...., myPayload, ...)

then we're currently forced to allocate another buffer to write the frame header. If we can tweak the allocator to reserve a certain number of bytes at the head of the returned buffer then we won't have to do two allocations.